### PR TITLE
Separate face matching from _compute_facial_adjacency_from_vertices

### PR DESCRIPTION
**Pull request by majosm**
Pulled out of #204 to declutter.

:warning: Remove `_FlatFacialAdjacencyData` again before merging. :warning:

### CODE REVIEW

File: meshmode/mesh/__init__.py [hunk 1]
```diff
@@ -1048,41 +1048,68 @@ def _boundary_tag_bit(boundary_tags, btag_to_index, boundary_tag):
...
 # {{{ vertex-based facial adjacency
 
-class _FlatFacialAdjacencyData:
+class _FaceIDs:
     """
-    Data structure for intermediate storage of facial adjacency data. Each attribute
-    is a :class:`numpy.ndarray` containing data for each stored face and its
-    adjacent neighbor.
+    Data structure for storage of a list of face identifiers (group, element, face).
+    Each attribute is a :class:`numpy.ndarray`.
```
Comment (inducer) [inline meshmode/mesh/__init__.py:1054]: Document shape of arrays.

File: meshmode/mesh/__init__.py [hunk 2]
```diff
@@ -1048,41 +1048,68 @@ def _boundary_tag_bit(boundary_tags, btag_to_index, boundary_tag):
...
+def _concatenate_face_ids(face_ids_list):
+    return _FaceIDs(
+        groups=np.concatenate([ids.groups for ids in face_ids_list]),
+        elements=np.concatenate([ids.elements for ids in face_ids_list]),
+        faces=np.concatenate([ids.faces for ids in face_ids_list]))
 
-        The group containing the adjacent element, or -1 if the face is not shared.
 
-    .. attribute:: neighbors
+def _match_faces_by_vertices(groups, face_ids, vertex_index_map_func=None):
```
Comment (inducer) [inline meshmode/mesh/__init__.py:1081]: Document return value.

File: meshmode/mesh/__init__.py [hunk 3]
```diff
@@ -1091,56 +1118,46 @@ def _compute_facial_adjacency_from_vertices(groups, boundary_tags,
...
-    for igrp, grp in enumerate(groups):
-        for fid, ref_fvi in enumerate(grp.face_vertex_indices()):
-            face_nr_base = face_nr_bases[igrp, fid]
-            grp_fvi = grp.vertex_indices[:, ref_fvi]
-            istart = face_nr_base
-            iend = face_nr_base + grp.nelements
-            face_vertex_indices[:len(ref_fvi), istart:iend] = grp_fvi.T
-            face_ids[0, istart:iend] = igrp
-            face_ids[1, istart:iend] = fid
+    matches = _match_faces_by_vertices(groups, face_ids)
```
Comment (inducer) [inline meshmode/mesh/__init__.py:1138]: `matches`: better variable name

File: meshmode/mesh/__init__.py [hunk 4]
```diff
@@ -1091,56 +1118,46 @@ def _compute_facial_adjacency_from_vertices(groups, boundary_tags,
...
-    adjacent_face_indices[matching_faces[0]] = matching_faces[1]
-    adjacent_face_indices[matching_faces[1]] = matching_faces[0]
+    # Get ((grp#, elem#, face#), (neighbor grp#, neighbor elem#, neighbor face#))
+    # for every face (both ways)
+
+    matches_both_ways = np.stack((
+        np.concatenate((matches[0, :], matches[1, :])),
+        np.concatenate((matches[1, :], matches[0, :]))))
+    order = np.lexsort((matches_both_ways[1, :], matches_both_ways[0, :]))
+    sorted_matches = matches_both_ways[:, order]
```
Comment (inducer) [inline meshmode/mesh/__init__.py:1150]: Explain reasoning for sort.

**Review by inducer — COMMENTED** [review]
Thanks! Some review bits that came up during our meeting.

**Comment by inducer** [comment]
Unsubscribing... @-mention or request review once it's ready for a look or needs attention.

**Comment by majosm** [comment]
Ready for a look @inducer.

I temporarily restored `_FlatFacialAdjacencyData` to make the diff a little less nasty. We'll have to remove it again before merging.

File: meshmode/mesh/__init__.py [hunk 5]
```diff
@@ -1085,62 +1085,131 @@ def __init__(self, nfaces, element_id_dtype, face_id_dtype):
         self.neighbor_faces = np.empty(nfaces, dtype=face_id_dtype)
 
 
+class _FaceIDs:
+    """
+    Data structure for storage of a list of face identifiers (group, element, face).
+    Each attribute is a :class:`numpy.ndarray` of shape (nfaces,).
```
Comment (inducer) [inline meshmode/mesh/__init__.py:1091]:
```suggestion
    Each attribute is a :class:`numpy.ndarray` of shape ``(nfaces,)``.
```

File: meshmode/mesh/__init__.py [hunk 6]
```diff
@@ -1085,62 +1085,131 @@ def __init__(self, nfaces, element_id_dtype, face_id_dtype):
...
+    return _FaceIDs(
+        groups=np.concatenate([ids.groups for ids in face_ids_list]),
+        elements=np.concatenate([ids.elements for ids in face_ids_list]),
+        faces=np.concatenate([ids.faces for ids in face_ids_list]))
+
+
+def _match_faces_by_vertices(groups, face_ids, vertex_index_map_func=None):
+    """
+    Return matching faces in *face_ids*, where two faces match if they have the
+    same vertices. Expressed as a :class:`np.ndarray` of shape (2,nmatches) of
```
Comment (inducer) [inline meshmode/mesh/__init__.py:1121]:
```suggestion
    same vertices. Expressed as a :class:`np.ndarray` of shape ``(2, nmatches)`` of
```

File: meshmode/mesh/__init__.py [hunk 7]
```diff
@@ -1085,62 +1085,131 @@ def __init__(self, nfaces, element_id_dtype, face_id_dtype):
...
+def _concatenate_face_ids(face_ids_list):
+    return _FaceIDs(
+        groups=np.concatenate([ids.groups for ids in face_ids_list]),
+        elements=np.concatenate([ids.elements for ids in face_ids_list]),
+        faces=np.concatenate([ids.faces for ids in face_ids_list]))
+
+
+def _match_faces_by_vertices(groups, face_ids, vertex_index_map_func=None):
+    """
+    Return matching faces in *face_ids*, where two faces match if they have the
```
Comment (inducer) [inline meshmode/mesh/__init__.py:1120]: > Return matching faces

In what form?

File: meshmode/mesh/__init__.py [hunk 8]
```diff
@@ -1085,62 +1085,131 @@ def __init__(self, nfaces, element_id_dtype, face_id_dtype):
...
+    same vertices. Expressed as a :class:`np.ndarray` of shape (2,nmatches) of
+    indices into *face_ids*. *vertex_index_map_func* is used to map vertices to
+    other vertices; it must accept (possibly multidimensional) numpy arrays of
+    vertex indices.
```
Comment (inducer) [inline meshmode/mesh/__init__.py:1124]: Document arguments. Specifically, document the signature and return type (also acceptable shapes) of `vertex_index_map_func`.

Also document the return type and meaning of returned data from `_match_faces_by_vertices`.

File: meshmode/mesh/__init__.py [hunk 9]
```diff
@@ -1085,62 +1085,131 @@ def __init__(self, nfaces, element_id_dtype, face_id_dtype):
...
+    for igrp, grp in enumerate(groups):
+        for fid, ref_fvi in enumerate(grp.face_vertex_indices()):
+            indices, = np.where((face_ids.groups == igrp) & (face_ids.faces == fid))
+            grp_fvi = grp.vertex_indices[face_ids.elements[indices], :][:, ref_fvi]
+            face_vertex_indices[:len(ref_fvi), indices] = (
+                vertex_index_map_func(grp_fvi).T)
+
+    # Lexicographically sort the face vertex indices, then diff the result to find
+    # faces with the same vertices
+    face_vertex_indices_increasing = np.sort(face_vertex_indices, axis=0)
```
Comment (inducer) [inline meshmode/mesh/__init__.py:1151]:
```suggestion
    
    # normalize vertex-based "face identifiers" by sorting
    face_vertex_indices_increasing = np.sort(face_vertex_indices, axis=0)
```

File: meshmode/mesh/__init__.py [hunk 10]
```diff
@@ -1085,62 +1085,131 @@ def __init__(self, nfaces, element_id_dtype, face_id_dtype):
...
+
+    face_index_pairs_both_ways = np.stack((
+        np.concatenate((
+            face_index_pairs[0, :],
+            face_index_pairs[1, :])),
+        np.concatenate((
+            face_index_pairs[1, :],
+            face_index_pairs[0, :]))))
+    # Sort by group, then neighbor group. Sort face indices because face_ids is
+    # already ordered by group
```
Comment (inducer) [inline meshmode/mesh/__init__.py:1198]:
```suggestion
    # Accomplish a sort by group, then neighbor group. This is done by sorting by the indices
    # in face_ids. Realize that those are already ordered by group by construction.
```

**Review by inducer — COMMENTED** [review]
Thanks! This looks great functionally. Just some nitpicking around docs and explanation, so that (hopefully) reading this becomes easier, then this is ready to go.

**Comment by majosm** [comment]
> > Remove `_FlatFacialAdjacencyData` again before merging.
> 
> Don't see that anywhere? What's this referring to?

It's right above `_FaceIDs` in `meshmode/mesh/__init__.py`. I removed it because it isn't needed anymore, but then I added it back in fe50f623b4082 because removing it made the diff hard to follow.

**Review by inducer — APPROVED** [review]
Thanks! This LGTM now. I'll wait for you to remove `_FlatFacialAdjacencyData`. Let me know if/when it's ready to merge from your end.

**Comment by majosm** [comment]
Good to go @inducer.

**Comment by inducer** [comment]
Thanks! (Hope you're OK with me squashing this... if not, please tell me, although I won't be able to do much to fix it.)
